lib/common/url.c

```diff
@@ -119,40 +119,42 @@ static h2o_iovec_t rebuild_path(h2o_mem_pool_t *pool, const char *src, size_t sr
 
 h2o_iovec_t h2o_url_normalize_path(h2o_mem_pool_t *pool, const char *path, size_t len, size_t *query_at, size_t **norm_indexes)
 {
-    const char *p = path, *end = path + len;
     h2o_iovec_t ret;
 
     *query_at = SIZE_MAX;
     *norm_indexes = NULL;
 
-    if (len == 0) {
+    if (len == 0 || (len == 1 && path[0] == '/')) {
         ret = h2o_iovec_init("/", 1);
         return ret;
     }
 
     if (path[0] != '/')
         goto Rewrite;
 
-    for (; p + 1 < end; ++p) {
-        if ((p[0] == '/' && p[1] == '.') || p[0] == '%') {
-            /* detect false positives as well */
-            goto Rewrite;
-        } else if (p[0] == '?') {
-            *query_at = p - path;
-            goto Return;
+    {
+        const char *p = path, *end = path + len;
+        for (; p + 1 < end; ++p) {
+            if ((p[0] == '/' && p[1] == '.') || p[0] == '%') {
+                /* detect false positives as well */
+                goto Rewrite;
+            } else if (p[0] == '?') {
+                *query_at = p - path;
+                goto Return;
+            }
         }
-    }
-    for (; p < end; ++p) {
-        if (p[0] == '?') {
-            *query_at = p - path;
-            goto Return;
+        for (; p < end; ++p) {
+            if (p[0] == '?') {
+                *query_at = p - path;
+                goto Return;
+            }
         }
-    }
 
-Return:
-    ret.base = (char *)path;
-    ret.len = p - path;
-    return ret;
+    Return:
```

Comment:
Generally speaking, we only allow jumps to the end of the function. This change is an violation to that rule.

I get the idea of limiting the scope of variables, but IMO it is more important to not have code jumping to arbitrary locations.

Reply:
Agreed. Reverting.